File: 剑指offer/15-二进制中1的个数P100/Solution.cpp
// Author : lihongyu
// Time   : 2019-11-02

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
     int NumberOf1(int n) {
         int count = 0;
         int flag = 1;

         while (flag){
             if(n & flag)
                 count++;
             flag = flag << 1;
         }
         return count;
     }
     
     int NumberOf12(int n) {
         int count = 0;
         
         while (n){
            count++;
            n = (n-1) & n;
         }
         return count;
     }
};

int main() {
    cout<<Solution().NumberOf12(9)<<endl;
    return 0;
}

File: 剑指offer/15-二进制中1的个数P100/Solution.java
// Author : lihongyu
// Time   : 2020-2-24
// leetcode-191 二进制中1的个数
// 思路：位运算
// 时间复杂度: O(n)
// 空间复杂度: O(n)
public class Solution {
    // you need to treat n as an unsigned value
    public int hammingWeight(int n) {
        int count = 0;
         
         while (n != 0){
            count++;
            n = (n-1) & n;
         }
         return count;
    }
}


// TODO: